src/test/java/com/microsoft/azure/relay/HybridConnectionListenerTest.java

```diff
@@ -29,45 +33,135 @@
+				assertFalse("client connection should be closed implicitly when rendezvous connection closes", clientConnection.isOpen());
+			});
+		});
+		
+		CompletableFuture.allOf(clientClosing, rendezvousClosing).whenComplete(($void, ex) -> {
+			if (ex != null) {
+				fail(ex.getMessage());
+			}	
+			listener.close();
+		}).join();
 	}
 	
 	@Test
-	public void acceptWebSocketConnectionTest() {
-		CompletableFuture<Boolean> checkSocketConnectionTask = new CompletableFuture<Boolean>();
-		CompletableFuture<HybridConnectionChannel> conn = listener.acceptConnectionAsync();
-		CompletableFuture<HybridConnectionChannel> clientConnectionTask = client.createConnectionAsync();
-		conn.thenAccept((connection) -> {
-			checkSocketConnectionTask.complete(true);
-			clientConnectionTask.thenAccept(clientConnection -> clientConnection.closeAsync());
+	public void acceptHandlerTest() throws InterruptedException, ExecutionException, TimeoutException {
+		HybridConnectionListener listener = new HybridConnectionListener(CONNECTION_URI, tokenProvider);
+		AtomicInteger handlerExecuted = new AtomicInteger(0);
+		listener.openAsync().join();
+		
+		// Handler accepts the connection
+		listener.setAcceptHandler(context -> {
+			handlerExecuted.incrementAndGet();
+			return true;
 		});
-		assertTrue("Listener failed to accept connections from sender in webSocket mode.", checkSocketConnectionTask.join());
+		
+		client.createConnectionAsync().whenComplete((clientConnection, ex) -> {
+			assertTrue("clientConnection should have been accepted", clientConnection != null && clientConnection.isOpen());
+		});
+		listener.acceptConnectionAsync().whenComplete((rendezvousConnection, ex) -> {
+			assertTrue("rendezvousConnection should have been accepted", rendezvousConnection != null && rendezvousConnection.isOpen());
+			rendezvousConnection.closeAsync();
+		}).join();
+		
+		// Handler rejects the connection
+		AtomicReference<RelayedHttpListenerContext> httpContext = new AtomicReference<RelayedHttpListenerContext>();
+		listener.setAcceptHandler(context -> {
+			httpContext.set(context);
+			handlerExecuted.incrementAndGet();
+			return false;
+		});
+		CompletableFuture<HybridConnectionChannel> acceptConnectionTask = listener.acceptConnectionAsync();
+		
+		client.createConnectionAsync().handle((clientConnection, ex) -> {
+			assertNull("clientConnection should not have been accepted", clientConnection);
+			assertNotNull("Rejection should have thrown.", ex);
+
+			RelayedHttpListenerResponse response = httpContext.get().getResponse();
+			assertEquals("The rejection response code is incorrect", HttpStatus.BAD_REQUEST_400, response.getStatusCode());
+			assertEquals("The rejection response description is incorrect", "Rejected by user code", response.getStatusDescription());
+			assertEquals("Both handlers should have been run", 2, handlerExecuted.get());
+			
+			listener.close();
```

Comment:
>listener [](start = 3, length = 8)

If the test fails prior to this point does the listener stay connected?

Reply:
I think this is fixed now, right?

---
In reply to: [261785842](https://github.com/Azure/azure-relay-java/pull/28#discussion_r261785842) [](ancestors = 261785842)

Reply:
it's fixed

---
In reply to: [262252931](https://github.com/Azure/azure-relay-java/pull/28#discussion_r262252931) [](ancestors = 262252931,261785842)